quint/src/quintVerifier.ts

```diff
@@ -311,11 +315,141 @@ async function retryWithTimeout<T>(f: () => Promise<T>): Promise<VerifyResult<T>
   ])
 }
 
-// Try to establish a connection to the Apalache server
-//
-// A successful connection procudes an `Apalache` object.
-async function connect(cmdExecutor: AsyncCmdExecutor): Promise<VerifyResult<Apalache>> {
-  return retryWithTimeout(() => cmdExecutor.ping()).then(response => response.map(_ => apalache(cmdExecutor)))
+/**
+ * Connect to the Apalache server, and verify the connection by pinging the server for up to 5 seconds.
+ *
+ * @returns A promise resolving to:
+ *    - a `right<Apalache>` if the connection is successful, or
+ *    - a `left<VerifyError>` if either the connection attempt is unsuccessful or pinging timed out.
+ */
+async function tryConnect(): Promise<VerifyResult<Apalache>> {
+  // Attempt to load proto definition:
+  // - if APALACHE_DIST is set, from the Apalache distribution
+  // - otherwise, via gRPC reflection
+  const protoDefResult: VerifyResult<proto.PackageDefinition> = process.env.APALACHE_DIST
+    ? findApalacheDistribution().chain(loadProtoDefViaDistribution)
+    : await loadProtoDefViaReflection()
+  // Load gRPC client
+  const maybeCmdExecutor = protoDefResult.map(loadGrpcClient)
+  const pingResult = await maybeCmdExecutor.asyncChain(cmdExecutor =>
+    // Try to ping the server, with a timeout
+    retryWithTimeout(() => cmdExecutor.ping())
+  )
+  return pingResult.chain(_ => maybeCmdExecutor.map(apalache))
```

Comment:
I understand this can be chained, but I'm skeptical of chaining `Either`s to achieve logically distinct things (in this case, loading the gRPC client vs. pinging).

I don't even know where to put comments in this case.

Reply:
Yeah, it's maybe just a question of style. IMO they are *not* logically distinct tho in the sense that we cannot ping unless we have the grpc client loaded. The chaining us justified, imo, by the logical dependency and the fact that that one way or another we are sequencing the error-prone computations.